### PR TITLE
Cherry-pick persist_stats_cf from rocksdb to tikv-3.0

### PR DESCRIPTION
**Pull request by GlitterIsMe**
Cherry-pick from commit of RocksDB:
c4f5d0aa152a0f0d8d603c91cbc2999993562c17
671d15cbdd3839acb54cb21a2aa82efca4917155

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
Thanks for the hard work! I'll double check the change later today.

File: db/db_impl.cc [hunk 1]
```diff
@@ -620,19 +634,169 @@ void DBImpl::PrintStatistics() {
...
+    }
+    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
+    if (stats_persist_period_sec > 0) {
+      if (!thread_persist_stats_) {
+        thread_persist_stats_.reset(new rocksdb::RepeatableThread(
+            [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
+            static_cast<uint64_t>(stats_persist_period_sec) * kMicrosInSecond));
+      }
+    }
+    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
```
Comment (yiwu-arbug) [inline db/db_impl.cc]: duplicated code

File: db/db_impl.h [hunk 2]
```diff
@@ -47,6 +47,7 @@
 #include "rocksdb/env.h"
 #include "rocksdb/memtablerep.h"
 #include "rocksdb/status.h"
+#include "rocksdb/stats_history.h"
```
Comment (yiwu-arbug) [inline db/db_impl.h]: this line seems not in original patch.

File: include/rocksdb/db.h [hunk 3]
```diff
@@ -1200,6 +1202,32 @@ class DB {
...
+  // Column families created by the primary after the secondary instance starts
+  // will be ignored unless the secondary instance closes and restarts with the
+  // newly created column families.
+  // Column families that exist before secondary instance starts and dropped by
+  // the primary afterwards will be marked as dropped. However, as long as the
+  // secondary instance does not delete the corresponding column family
+  // handles, the data of the column family is still accessible to the
+  // secondary.
+  // TODO: we will support WAL tailing soon.
+  virtual Status TryCatchUpWithPrimary() {
```
Comment (yiwu-arbug) [inline include/rocksdb/db.h]: `TryCatchUpWithPrimary` is from other patches. remove.

File: options/db_options.cc [hunk 4]
```diff
@@ -217,6 +218,12 @@ void ImmutableDBOptions::Dump(Logger* log) const {
                    two_write_queues);
   ROCKS_LOG_HEADER(log, "            Options.manual_wal_flush: %d",
                    manual_wal_flush);
+  ROCKS_LOG_HEADER(log, "            Options.atomic_flush: %d", atomic_flush);
+  ROCKS_LOG_HEADER(log,
```
Comment (yiwu-arbug) [inline options/db_options.cc]: `avoid_unnecessary_blocking_io` is not part of the change. remove.

File: options/db_options.h [hunk 5]
```diff
@@ -79,6 +79,8 @@ struct ImmutableDBOptions {
   bool two_write_queues;
   bool manual_wal_flush;
   bool atomic_flush;
+  bool avoid_unnecessary_blocking_io;
```
Comment (yiwu-arbug) [inline options/db_options.h]: remove `avoid_unnecessary_blocking_io`

**Comment by yiwu-arbug** [comment]
Fix the above and we are ready to go.

File: tools/db_bench_tool.cc [hunk 6]
```diff
@@ -1045,6 +1045,17 @@ DEFINE_bool(identity_as_first_hash, false, "the first hash function of cuckoo "
...
               "Gap between printing stats to log in seconds");
+DEFINE_uint64(stats_persist_period_sec,
+              rocksdb::Options().stats_persist_period_sec,
+              "Gap between persisting stats in seconds");
+DEFINE_bool(persist_stats_to_disk, rocksdb::Options().persist_stats_to_disk,
+            "whether to persist stats to disk");
+DEFINE_uint64(stats_history_buffer_size,
+              rocksdb::Options().stats_history_buffer_size,
+              "Max number of stats snapshots to keep in memory");
+DEFINE_int64(multiread_stride, 0,
```
Comment (Connor1996) [inline tools/db_bench_tool.cc]: irrelevant change?

Comment (GlitterIsMe) [reply tools/db_bench_tool.cc]: this is added by this pr, I think it maybe useful...

**Comment by yiwu-arbug** [comment]
Will make another attempt to cherry-pick those changes later on. Closing for now.
